### ISSUE TITLE
SymPy does not evaluate integrals of exponentials with symbolic parameter and limit

### ISSUE THREAD

**Issue by fredrik-johansson** [sympy/sympy#13312]
This ought to be easy:

    >>> var("a b t")
    (a, b, t)
    >>> integrate(exp(-a*t), (t,b,oo))
    Integral(exp(-a*t), (t, b, oo))

**Comment by jksuom** [comment]
Apparently, the limit should be implemented:
```
>>>limit(exp(-a*t), t, oo)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sympy/series/limits.py", line 49, in limit
    return Limit(e, z, z0, dir).doit(deep=False)
  File "sympy/series/limits.py", line 193, in doit
    raise NotImplementedError()
NotImplementedError
```

**Comment by wsdookadr** [comment]
> `>>>limit(exp(-a*t), t, oo)`

Isn't this one 0 for positive `a` and oo for negative `a`?
Individually, each of those cases work. But without restricting a to
positive/negative, the result should be a piecewise function.

**Comment by skirpichev** [comment]
> Isn't this one 0 for positive a and ∞ for negative a

And 1 for a == 0...

This issue was fixed in the Diofant, but parametric limits are handled not quite well.  See diofant/diofant#425.  E.g., ``limit(exp(-a*t), t, oo)`` should be ``Piecewise[(1, Eq(a, 0)), (exp(-sign(a)*oo), True))``.

**Comment by jksuom** [comment]
Maybe `re(a)` should be considered if `a` is not known to be real. And the case of purely imaginary `a`

**Comment by wsdookadr** [comment]
@jksuom 
I am going to quote [a comment](https://github.com/sympy/sympy/blob/94a24240ef3b6016afe8e044f62a5ef99f5fc491/sympy/series/gruntz.py#L508) from gruntz.py

> For limits of complex functions, the algorithm would have to be
> improved, or just find limits of Re and Im components separately.

This is one hint that the current implementation is intended for real functions.
Another one would be that the paper itself(just from skimming through it)
seems to deal with real-valued functions of real variables.

**Comment by jksuom** [comment]
Gruntz's algorithm is based on the theory of Hardy fields whose elements are equivalence classes of real functions defined and not changing sign on real intervals `(a, oo)`. In SymPy, some effort is made to extend this to complex functions but with limited success. Another problem is that the theory cannot handle functions of variable sign.

**Comment by wsdookadr** [comment]
@jksuom The part about handling variable sign could be approached by first identifying all constants, enumerating all their possible sign combinations, then handling each sign combination with the existing logic, and at the end, building a piecewise function with one branch per sign combination. In our case, there would actually be three such branches. Would this approach be useful?

**Comment by skirpichev** [comment]
> The part about handling variable sign could be approached by first identifying all constants, enumerating all their possible sign combinations...

This will not work :)  Before invocation of the algorithm - you can't know which combination of constants you would like to test.   In this case - it's just ``a``.  But for another one-parametric function it could be a more generic expression like ``a - 1`` or ``a**2 - 1``.

Yet you could, in principle, catch mentioned above exceptions (like "Result depends on the sign of ...") - and then re-run algorithm with added assumptions for parameters (e.g. ``a > 0``), building a piecewise function from all such cases at the end.  But 1) you should prove first that this extension of the Gruntz algorithm - is an algorithm itself, not a heuristics that **may** work.  Will be number of different cases finite?  And 2) unfortunately, so-called "old" assumptions in the SymPy don't allow to specify generic assumptions for non-symbols, i.e. for expressions like ``a - 1``.

**Comment by wsdookadr** [comment]
@skirpichev Ok, but then how about a change of variables `θ = a-1` and `φ = a**2 - 1`. Then do what we talked about above with the enumeration(but with expressions with `θ,φ`). And at the end, solve `φ > 0` and `φ < 0` in `a` and find out what the branches will be (for example `φ` will create two branches, a>1 and a<1).

**Comment by skirpichev** [comment]
> Ok, but then how about a change of variables θ = a-1 and ...

Yeah, place assumptions on symbol θ and substitute a=θ+1 back into expression...  That could solve 2)-nd issue, more technical one.

**Comment by asmeurer** [comment]
The error from Gruntz itself is 

```
Traceback (most recent call last):
  File "./sympy/series/limits.py", line 180, in doit
    r = gruntz(e, z, z0, dir)
  File "./sympy/series/gruntz.py", line 648, in gruntz
    r = limitinf(e, z)
  File "./sympy/series/gruntz.py", line 428, in limitinf
    c0, e0 = mrv_leadterm(e, x)
  File "./sympy/series/gruntz.py", line 512, in mrv_leadterm
    f, logw = rewrite(exps, Omega, x, w)
  File "./sympy/series/gruntz.py", line 584, in rewrite
    raise NotImplementedError('Result depends on the sign of %s' % sig)
NotImplementedError: Result depends on the sign of -sign(a)
```

This "result depends on sign(expr)" error comes up a lot. Shouldn't it be possible to split this into a 3-piecewise expression for the different values of sign(expr)?

**Comment by skirpichev** [comment]
In this case you have just one special value of the parameter, a=0.  In general, there may be more.  How many?  Can you prove that there will be only a finite number of such special values?  Say, if we just extend exp-log functions of one variable, x, by adding parameter a.

**Comment by asmeurer** [comment]
You don't need to solve for a in the cases. The Piecewise can just be `sign(expr) > 0`, `sign(expr) < 0`, `sign(expr) == 0`. Maybe I'm missing the point. Can you give an example where there would be a problem?

**Comment by wsdookadr** [comment]
@skirpichev 

> Can you prove that there will be only a finite number of such special values?

Well, let's say we were trying to determine `sign(sin(a))`. this wobbles between -1 and 1 an infinite number of times, and it has zeros for `a=k*pi` (with k being an integer) , so an infinite number of such special values. It does help that it's periodic though, and we can describe with precision when it's positive/negative/zero. 

There may be more complicated cases like that though.

Is this what you were refering to?

**Comment by skirpichev** [comment]
> Is this what you were refering to?

Yes.  Of course, no periodicity in a general case.

**Comment by asmeurer** [comment]
For `limit(exp(sin(a)*x), x, oo)` what's wrong with `Piecewise((oo, sin(a) > 0), (0, sin(a) < 0), (1, Eq(sin(a), 0))`?

Simplifying conditionals is always nice if we can do it, but it's not required, and it's a separate problem from limits.

**Comment by wsdookadr** [comment]
On a different note, let's look at what [Maxima](http://maxima.sourceforge.net/) says about this:

```
(%i6) limit(%e^(-a*x),x,inf);
Is a positive, negative or zero?
```

So the same thing, Maxima which has another implementation of Gruntz's algorithm raises the same exception.

**Comment by skirpichev** [comment]
> For limit(exp(sin(a)*x), x, oo) what's wrong with

Nothing. But lets take a random exp-log example.  I'll make this with Diofant.

```
In [1]: from diofant.series.gruntz import limitinf
In [2]: x = Symbol('x', real=True, positive=True)
In [3]: a = Symbol('a')
In [4]: p = Symbol('p', positive=True)
In [5]: limitinf((exp(x + log(a + exp(-x/a))) - exp(x**a)), x)
[...]
NotImplementedError: Result depends on the sign of -sign(a)
```
We can substitute a positive symbol to examine one branch of our future Piecewise expression, but...
```
In [6]: limitinf((exp(x + log(a + exp(-x/a))) - exp(x**a)).subs(a, p), x)
[...]
NotImplementedError: Result depends on the sign of sign(-p + 1)
```
And only after second substitution we got answer:
```
In [7]: limitinf((exp(x + log(a + exp(-x/a))) - exp(x**a)).subs(a, 1 + p), x)
Out[7]: -∞
```

As you see, on practice we don't have an analytic expression of special values from beginning, as in your example.  Also, this time we discover "special values" - twice.  But in general, I'm worry that number of such events could be infinite...

> What do you think of this approach to not fully solve the branch conditions

@wsdookadr  it's impossible with so-called "old" assumption system in sympy.  E.g. you can make symbol positive with ``Symbol('p', positive=True)``, but you can't add assumption on ``Add(p, 1)``.

**Comment by asmeurer** [comment]
Let's try to implement the Piecewise logic, and see if it leads to infinite recursion. At worst may need to limit the depth of the Piecewise. And probably also some cycle detection. 

I also noticed that things can be dependent on one another. If you dig in the code, it depends on `sign(a)`, `-sign(a)`, `sign(-a - 1)`, `sign(1/a)`, and `sign((a - 1)/a)`, which are obviously not all independent. So some heuristics will need to be implemented to avoid all 2^n possible splits. Fortunately, even if a split is redundant or impossible, that is mathematically handled by the definition of a Piecewise. So heuristics are fine. At worst you might get an odd, but at least not mathematically incorrect answers in some cases. 

I would try to implement the splits at a low level, in gruntz itself. Just doing subs on the original expression could lead to problems because of the way subs tries to be smart, and because substituting in an expression can lead to cancellation.

**Comment by JohnStone2017** [comment]
I wish I could offer to help fix this, but it is over my head. In the meantime, I am having to consider piping strings to and from Maxima or Reduce from Python in order to do symbolic integration. This type of of integral is something all other CAS can do, no?

**Comment by JohnStone2017** [comment]
This looks fixed; at least it solves an integral it couldn't several months ago. Thanks much; replacing Mathematica with Sympy is my goal, and this goes a long way.

**Comment by Davide-sd** [comment]
Just happen to run this limit on Sympy 1.5.1:

```
import sympy as sp
x = sp.symbols("x", real=True)
alpha = sp.symbols("alpha", real=True, zero=True)
sp.limit(sp.exp(-alpha * x), x, sp.oo)
```

Output: `exp(-oo*alpha)`

Was expecting 1.

**Comment by asmeurer** [comment]
Setting zero=True on a symbol won't always work, compared to just using `0` directly. The issue is that even with `alpha = Symbol('alpha', real=True)`, `exp(-oo*alpha)` is not valid for alpha=0 (at least it would give nan instead of an incorrect result). Once we have limit that can return Piecewise, it might be better to split out the alpha=0 case. Actually, a three-way piecewise with 0, 1, and oo would be a much better result than a symbolic one.

**Comment by FO2020** [comment]
Hei,I want to jump in with another example which give the same error.
```x,y1,y2=sp.symbols('x,y1,y2',real=True,nonnegative=True)
a,b,c1,c2=sp.symbols('a,b,c1,c2',real=True,positive=True)
delta1,delta2,psi,phi=sp.symbols('delta1,delta2,psi,phi',real=True,positive=True)

y1sol=y10*exp(t*(delta1 - x_m))
[limit(y1sol,t,0),limit(y1sol,t,oo)]
```
Error message:
`---------------------------------------------------------------------------
NotImplementedError                       Traceback (most recent call last)
<ipython-input-164-88cb6f289e95> in <module>
      1 sp.assumptions.assume.assuming(delta1-x_m<0)
----> 2 [limit(y1sol,t,0),limit(y1sol,t,oo)]

C:\ProgramData\Anaconda3\lib\site-packages\sympy\series\limits.py in limit(e, z, z0, dir)
     69     """
     70 
---> 71     return Limit(e, z, z0, dir).doit(deep=False)
     72 
     73 

C:\ProgramData\Anaconda3\lib\site-packages\sympy\series\limits.py in doit(self, **hints)
    259                             % (l, r))
    260             else:
--> 261                 r = gruntz(e, z, z0, dir)
    262             if r is S.NaN or l is S.NaN:
    263                 raise PoleError()

C:\ProgramData\Anaconda3\lib\site-packages\sympy\series\gruntz.py in gruntz(e, z, z0, dir)
    669 
    670     try:
--> 671         r = limitinf(e0, z)
    672     except ValueError:
    673         r = limitinf(e0, z, leadsimp=True)

C:\ProgramData\Anaconda3\lib\site-packages\sympy\series\gruntz.py in limitinf(e, x, leadsimp)
    433         e = e.subs(x, p)
    434         x = p
--> 435     c0, e0 = mrv_leadterm(e, x)
    436     sig = sign(e0, x)
    437     if sig == 1:

C:\ProgramData\Anaconda3\lib\site-packages\sympy\series\gruntz.py in mrv_leadterm(e, x)
    521     #
    522     w = Dummy("w", real=True, positive=True, finite=True)
--> 523     f, logw = rewrite(exps, Omega, x, w)
    524     series = calculate_series(f, w, logx=logw)
    525     return series.leadterm(w)

C:\ProgramData\Anaconda3\lib\site-packages\sympy\series\gruntz.py in rewrite(e, Omega, x, wsym)
    593         sig = sign(g.args[0], x)
    594         if sig != 1 and sig != -1:
--> 595             raise NotImplementedError('Result depends on the sign of %s' % sig)
    596     if sig == 1:
    597         wsym = 1/wsym  # if g goes to oo, substitute 1/w

NotImplementedError: Result depends on the sign of sign(delta1 - x_m)``

```
Why not giving the limit in the two cases instead of the error message? I wonder also if we can set assumption `delta1<x_m` in sympy?

**Comment by asmeurer** [comment]
Presently there's no way to make that assumption in a way that will be seen by `limit`. The only fix is to replace `delta1 - x_m` with a single symbol that has the `positive=True` assumption for the calculation.

**Comment by FO2020** [comment]
@asmeurer Thanks.That’s what  I did actually.It will be perfect to add something like that in the future.

**Comment by asmeurer** [comment]
Ideally one could use `with assuming(Q.positive(delta1 - x_m))`, but the problem is that the limits code doesn't read the new assumptions, so it does nothing. Once this issue is fixed and limit returns a Piecewise, you could also use refine() to simplify the expression.
